bee-pow/src/providers/miner.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Contains a nonce provider that mine nonces.
+
+use crate::providers::{NonceProvider, NonceProviderBuilder};
+
+use bee_crypto::ternary::{
+    sponge::{BatchHasher, CurlPRounds, BATCH_SIZE},
+    HASH_LENGTH,
+};
+use bee_ternary::{b1t6, Btrit, T1B1Buf, TritBuf};
+
+use crypto::hashes::{blake2b::Blake2b256, Digest};
+use thiserror::Error;
+
+use std::{
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread,
+};
+
+const DEFAULT_NUM_WORKERS: usize = 1;
+/// https://oeis.org/A002391
+const LN_3: f64 = 1.098_612_288_668_109_8;
+
+/// Errors occurring when computing nonces with the `Miner` nonce provider.
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("The worker has been cancelled.")]
+    Cancelled,
+}
+
+/// Builder for the `Miner` nonce provider.
+#[derive(Default)]
+pub struct MinerBuilder {
+    num_workers: Option<usize>,
+}
+
+impl MinerBuilder {
+    /// Sets the desired number of workers for the `Miner` nonce provider.
+    pub fn with_num_workers(mut self, num_workers: usize) -> Self {
+        self.num_workers = Some(num_workers);
+        self
+    }
+}
+
+impl NonceProviderBuilder for MinerBuilder {
+    type Provider = Miner;
+
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn finish(self) -> Miner {
+        Miner {
+            num_workers: self.num_workers.unwrap_or(DEFAULT_NUM_WORKERS),
+        }
+    }
+}
+
+/// A nonce provider that mine nonces.
+pub struct Miner {
+    num_workers: usize,
+}
+
+impl Miner {
+    fn worker(
+        done: Arc<AtomicBool>,
+        pow_digest: TritBuf<T1B1Buf>,
+        start_nonce: u64,
+        target_zeros: usize,
+    ) -> Result<u64, Error> {
+        assert!(target_zeros <= HASH_LENGTH);
```

Comment:
how likely is this assert to ever cause a panic? I wonder if we should make this `debug_assert`. If `target_zeros` is something the user controls in the config, then I think we should catch this there already instead of here.